### PR TITLE
Type annotations for tf dtypes

### PR DESCRIPTION
**Pull request by rahul-kamat**
These changes allow TensorFlow dtypes to have their own types by defining Python dtype classes for all tf dtypes. This will allow us to type annotate Tensors and dtypes provided as arguments to functions.
### Examples:
We can use this to annotate function arguments:
```python
def foo(dt: dtypes.Float32) # dt is expected to be tf.float32
```
```python
def foo(dt: tf.Tensor[dtypes.Float32]) # dt is expected to be a Tensor with dtype=tf.float32
```
Before: The type of tf.bool, tf.int16, tf.string, etc would be “DType” which meant it would not be possible to differentiate dtypes based on their types.

### Changes:
- 3b1c93b - create classes for each TF dtype to extend DType base class
- e2931cc - add stub file to give each TF dtype their respective type
- 0c1aed9 - annotations for dtype parameter for tf.saturate_cast
- 2292cee & d7f5e05 - type decorators
- cd36697 - annotations for dtype parameter for tf.constant

### CODE REVIEW

File: tensorflow/python/framework/dtypes.py
Comment (mdanatg) [inline tensorflow/python/framework/dtypes.py:241]: For backward compatibility, this needs to be written like so: `Half = Float16`. Same for all other dtypes that are just the alias of another.

File: tensorflow/python/framework/dtypes.py
Comment (mdanatg) [inline tensorflow/python/framework/dtypes.py:318]: half = float16

File: tensorflow/python/framework/dtypes.py
Comment (mdanatg) [inline tensorflow/python/framework/dtypes.py:362]: Let's create classes for these as well.

File: tensorflow/python/util/dispatch.py
Comment (mdanatg) [inline tensorflow/python/util/dispatch.py:48]: Let's make this comment more informative - this is a type var for decorators (or omit the comment altogether).

File: tensorflow/python/ops/math_ops.py
Comment (mdanatg) [inline tensorflow/python/ops/math_ops.py:940]: Let's annotate this file in a separate PR.

**Review by mdanatg — CHANGES_REQUESTED** [review]
Looks good. Let's set the op annotations aside for now, and fill them in later with types for all arguments and return values.

**Review by mdanatg — DISMISSED** [review]

**Review by mdanatg — DISMISSED** [review]

**Comment by gbaned** [comment]
@rahul-kamat Can you please address Ubuntu Sanity errors? Thanks!

**Review by mdanatg — APPROVED** [review]

**Review by gbaned — APPROVED** [review]

**Review by gbaned — APPROVED** [review]

**Comment by mdanatg** [comment]
@yashk2810 FYI

**Review by gbaned — APPROVED** [review]

**Comment by mdanatg** [comment]
@gbaned could you help get the import/copybara bit unstuck?

**Review by mdanatg — APPROVED** [review]

**Review by mdanatg — APPROVED** [review]

**Comment by mdanatg** [comment]
This should be ready to merge.

**Review by rthadur — APPROVED** [review]

**Comment by gbaned** [comment]
@rahul-kamat Can you please resolve conflicts? Thanks!

**Comment by tensorflowbutler** [comment]
It has been 15 days with no activity and the `awaiting response` label was assigned. Is this PR still valid? Assigning the `stalled` label. Please comment to reassure me that this is still being worked on.

**Comment by mdanatg** [comment]
The PR is still valid, but we've been delayed on the integration.

File: tensorflow/python/util/dispatch.py
Comment (joelberkeley) [inline tensorflow/python/util/dispatch.py:49]: should this be
```suggestion
F = TypeVar("F", bound=Callable[..., object])
```
perhaps there's no functional difference since its in a `TypeVar`, but it does eliminate an `Any`

Comment (joelberkeley) [reply tensorflow/python/util/dispatch.py:49]: (btw, hi, I'm interested in helping out with adding type hints to tensorflow, so getting involved where I can. If this comment isnt helpful, feel free to ignore)

Comment (mdanatg) [reply tensorflow/python/util/dispatch.py:49]: Changing to `object` sounds reasonable to me.

It looks like this PR is still blocked on some merge conflicts. If would be helpful to unblock it by resolving the conflicts and making sure the integration tests still pass. This might need cloning it into a new PR.

**Comment by rthadur** [comment]
I'm going to go ahead and close this PR, because it seems to have stalled. If you're still interested in pursing this (and responding to my comments), please feel free to reopen!

**Pull request closed without merging**
